implements StatefulSet update

kow3ns opened the pull request:
**What this PR does / why we need it**:
1. Implements rolling update for StatefulSets
2. Implements controller history for StatefulSets.
3. Makes StatefulSet status reporting consistent with DaemonSet and ReplicaSet.

https://github.com/kubernetes/features/issues/188

**Special notes for your reviewer**:

**Release note**:
```release-note
Implements rolling update for StatefulSets. Updates can be performed using the RollingUpdate, Paritioned, or OnDelete strategies. OnDelete implements the manual behavior from 1.6. status now tracks 
replicas, readyReplicas, currentReplicas, and updatedReplicas. The semantics of replicas is now consistent with DaemonSet and ReplicaSet, and readyReplicas has the semantics that replicas did prior to this release.
```

krmayankk commented on api/swagger-spec/apps_v1beta1.json:
```diff
@@ -6010,6 +6047,29 @@
+     "updatedReplicas": {
+      "type": "integer",
+      "format": "int32",
+      "description": "updatedReplicas is the number of replicas"
```
nit: at `updatedRevision`

krmayankk commented on api/swagger-spec/apps_v1beta1.json:
```diff
@@ -6010,6 +6047,29 @@
+     },
+     "updateRevision": {
+      "type": "string",
+      "description": "updateRevision is the revision the StatefulSet is attempting to roll out"
```
nit: comma after `revision` as in `updateRevision is the revision, the StatefulSet is ...`

krmayankk commented on docs/api-reference/apps/v1beta1/definitions.html:
```diff
@@ -6452,6 +6535,47 @@ <h3 id="_v1_configmapenvsource">v1.ConfigMapEnvSource</h3>
 <div class="sect2">
+<h3 id="_v1beta1_statefulsetupdatestrategy">v1beta1.StatefulSetUpdateStrategy</h3>
+<div class="paragraph">
+<p>StatefulSetUpdateStrategy indicates the strategy that the StatefulSet controller will use to perform updates. It includes any additional parameters necessary to preform the update for the indicated strategy.</p>
```
nit: `perform`

krmayankk commented on pkg/apis/apps/types.go:
```diff
@@ -60,6 +60,50 @@ const (
+	// PartitionStatefulSetStrategyType indicates that updates will only be
+	// applied to a partition of the StatefulSet. This is useful for canaries
+	// and phased roll outs. When a scale operation is performed with this
+	// strategy, new Pods will be created from the updated specification.
```
updated specification means updateRevision here, right ?

krmayankk commented on pkg/apis/apps/types.go:
```diff
@@ -60,6 +60,50 @@ const (
+	// tracking and ordered rolling restarts are disabled. Pods are recreated
+	// from the StatefulSetSpec when they are manually deleted. When a scale
+	// operation is performed with this strategy, new Pods will be created
+	// from the current specification.
```
current specification means currentRevision here , right ?

krmayankk commented on pkg/apis/apps/types.go:
```diff
@@ -60,6 +60,50 @@ const (
+	// from the StatefulSetSpec when they are manually deleted. When a scale
+	// operation is performed with this strategy, new Pods will be created
+	// from the current specification.
+	OnDeleteStatefulSetStrategyType = "OnDelete"
```
is this the default strategy ?

kow3ns replied:
no. you should look at defaults.go.

krmayankk commented on pkg/apis/apps/types.go:
```diff
@@ -60,6 +60,50 @@ const (
+type PartitionStatefulSetStrategy struct {
+	// Ordinal indicates the ordinal at which the StatefulSet should be
+	// partitioned.
+	Ordinal int32
```
Can this be omitted in the Partitioned case to assume a default of one ? In which case,this should be a pointer to distringuish between the default case .

krmayankk replied:
i think a default one one assumption will be helpful

kow3ns replied:
RollingUpdate is default

krmayankk replied:
the default partition size of 1 is the canary case. Most people would have this use case. Without it, it means specifying one extra param for partitioning in the most common case. It seems like a reasonable default with no surprises .

kow3ns replied:
I disagree that canaries are the most common case. I don't think specifying the fraction of a staged roll out is an undo burden.

smarterclayton replied:
I think we should require partition to be specified.  The default doesn't add anything for the machine operated use case which is going to be much more common

krmayankk commented on pkg/apis/apps/types.go:
```diff
@@ -109,6 +153,17 @@ type StatefulSetSpec struct {
+
+	// updateStrategy indicates the StatefulSetUpdateStrategy that will be
+	// employed to update Pods in the StatefulSet when a revision is made to
+	// Template or VolumeClaimsTemplate.
```
Can we remove the VolumeClaimsTemplate from here, since we dont support the update to it.

krmayankk commented on pkg/apis/apps/types.go:
```diff
@@ -119,6 +174,21 @@ type StatefulSetStatus struct {
+	// currentReplicas is the number of replicas at the CurrentRevision
+	CurrentReplicas int32
+
+	// updatedReplicas is the number of replicas
```
nit: at the updateRevision

krmayankk commented on pkg/apis/apps/v1beta1/conversion.go:
```diff
@@ -109,6 +109,13 @@ func Convert_v1beta1_StatefulSetSpec_To_apps_StatefulSetSpec(in *StatefulSetSpec
+	}
+	if in.RevisionHistoryLimit != nil {
+		out.RevisionHistoryLimit = new(int32)
+		*out.RevisionHistoryLimit = *in.RevisionHistoryLimit
```
and if its nil, set it to 2 ?

kow3ns replied:
defaults.go not conversions

krmayankk commented on pkg/apis/apps/v1beta1/conversion.go:
```diff
@@ -140,8 +147,37 @@ func Convert_apps_StatefulSetSpec_To_v1beta1_StatefulSetSpec(in *apps.StatefulSe
 	}
+	if in.RevisionHistoryLimit != nil {
+		out.RevisionHistoryLimit = new(int32)
+		*out.RevisionHistoryLimit = *in.RevisionHistoryLimit
```
in nil case, set it to 2

kow3ns replied:
no that is done in defaults not conversions

krmayankk commented on pkg/apis/apps/validation/validation_test.go:
```diff
@@ -202,6 +246,43 @@ func TestValidateStatefulSet(t *testing.T) {
+				UpdateStrategy:      apps.StatefulSetUpdateStrategy{Type: "foo"},
+			},
+		},
+		"empty parition": {
```
nit: partition

krmayankk commented on pkg/controller/statefulset/stateful_pod_control_test.go:
```diff
@@ -461,116 +459,6 @@ func TestStatefulPodControlDeleteFailure(t *testing.T) {
-		t.Errorf("Expected 0 events for successful status update %d", eventCount)
-	}
-}
-
```
has all this content moved elsewhere ?

kow3ns replied:
The status tests are in stateful_set_status_updater.go

krmayankk commented on pkg/controller/statefulset/stateful_set.go:
```diff
@@ -305,6 +308,32 @@ func (ssc *StatefulSetController) getPodsForStatefulSet(set *apps.StatefulSet, s
 
+// adoptOrphanRevisions adopts any orphaned ControllerRevisions matched by set's Selector.
+func (ssc *StatefulSetController) adoptOrphanRevisions(set *apps.StatefulSet) error {
+	revisions, err := ssc.control.ListRevisions(set)
```
is this listing of revisions based on the label on those revisions ?

kow3ns replied:
Yes the revisions are constructed with the StatefulSet's selector

krmayankk commented on pkg/controller/statefulset/stateful_set.go:
```diff
@@ -305,6 +308,32 @@ func (ssc *StatefulSetController) getPodsForStatefulSet(set *apps.StatefulSet, s
+		}
+	}
+	if hasOrphans {
+		fresh, err := ssc.kubeClient.AppsV1beta1().StatefulSets(set.Namespace).Get(set.Name, metav1.GetOptions{})
```
curious why cant we use the set.UID instead of querying again ?

kow3ns replied:
We have to refresh the Object to ensure that it hasn't been deleted ensure that we don't patch a ControllerRevision to be owned by a cached deleted/(deleted and recreated) StatefulSet.

krmayankk replied:
hmm how can you be sure of that ?What if its gets deleted after your query ?

kow3ns replied:
I suppose in theory it could. In practice we have seen issues where a stale cache causes invalid adoption. This pattern is used in all controllers for orphan Pods.

krmayankk commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -246,6 +255,23 @@ func newControllerRef(set *apps.StatefulSet) *metav1.OwnerReference {
+	if pod.Labels == nil {
+		pod.Labels = make(map[string]string)
+	}
+	pod.Labels[apps.StatefulSetRevisionLabel] = revision
```
check if revision is empty ?

kow3ns replied:
It can't be empty based on the call path. I don't see a need for a check.

krmayankk commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -255,6 +281,88 @@ func newStatefulSetPod(set *apps.StatefulSet, ordinal int) *v1.Pod {
+	return patch, err
+}
+
+// newRevision creates a new ControllerRevision containing a patch that can convert a StatefulSet to an equivalent state to set.
```
nit: the last part of the sentence is not clear `convert a StatefulSet to an equivalent state to set.`

krmayankk commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -255,6 +281,88 @@ func newStatefulSetPod(set *apps.StatefulSet, ordinal int) *v1.Pod {
+	return history.NewControllerRevision(set,
+		controllerKind,
+		selector,
+		runtime.RawExtension{Raw: patch},
```
this is confusing and at the same time interesting to me as a concept :-) I have not seen this pattern before, but it seems like what you are doing is keeping a strategic merge patch in each revision history , so that any time we want to restore the StatefulSet to that revision, you just use the ready to apply patch from the Raw field to restore it. I think this is worth some more comment

kow3ns replied:
I added commentary to that effect above the getPatch and newRevision methods.

krmayankk commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -255,6 +281,88 @@ func newStatefulSetPod(set *apps.StatefulSet, ordinal int) *v1.Pod {
+		return nil, err
+	}
+	clone := obj.(*apps.StatefulSet)
+	patched, err := strategicpatch.StrategicMergePatch([]byte(runtime.EncodeOrDie(patchCodec, clone)), revision.Data.Raw, clone)
```
check for err here ?

kow3ns replied:
I added a check under strategicpatch
If you mean check that the clone casts correctly, we generally do not do that when deep copying from using Scheme.

krmayankk commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -255,6 +281,88 @@ func newStatefulSetPod(set *apps.StatefulSet, ordinal int) *v1.Pod {
+	if err != nil {
+		return nil, err
+	}
+	return clone, err
```
return `clone, nil `

krmayankk commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -255,6 +281,88 @@ func newStatefulSetPod(set *apps.StatefulSet, ordinal int) *v1.Pod {
+func nextRevision(revisions []*apps.ControllerRevision) int64 {
+	if count := len(revisions); count <= 0 {
+		return 1
+	} else {
```
remove the else

krmayankk commented on pkg/controller/statefulset/stateful_set_utils_test.go:
```diff
@@ -287,6 +288,26 @@ func TestNewPodControllerRef(t *testing.T) {
+		t.Fatal(err)
+	}
+	set.Spec.Template.Spec.Containers[0].Name = "foo"
+	restored, err := applyRevision(set, revision)
```
nit: restoredSet ?

krmayankk commented on pkg/controller/statefulset/stateful_set_utils_test.go:
```diff
@@ -287,6 +288,26 @@ func TestNewPodControllerRef(t *testing.T) {
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !history.EqualRevision(revision, restoredRevision) {
```
do you ignore the revision number when comparing the revisions ?

kow3ns replied:
Yes the revisions are only compared by hash and by contents. Two revisions with different ".Revisions" can be equal using this comparison.

krmayankk commented on pkg/kubectl/history.go:
```diff
@@ -131,6 +133,40 @@ func (h *DeploymentHistoryViewer) ViewHistory(namespace, name string, revision i
+	if err != nil {
+		return "", fmt.Errorf("failed to retrieve statefulset %s", err)
+	}
+	revisions, err := h.c.Apps().ControllerRevisions(namespace).List(metav1.ListOptions{LabelSelector: sts.Spec.Selector.String()})
```
will this sometimes also report orphaned revisions ? To avoid that, we should further validate if the ownerReferences are for this set

krmayankk commented on pkg/kubectl/history.go:
```diff
@@ -131,6 +133,40 @@ func (h *DeploymentHistoryViewer) ViewHistory(namespace, name string, revision i
+		return "No rollout history found.", nil
+	}
+	revisionNumbers := make([]int64, len(revisions.Items))
+	for i := range revisions.Items {
```
is this loop necessary  ? just use revision.Items directly everywhere ?

kow3ns replied:
It is necessary to sort the slice of revisions for display.

krmayankk commented on staging/src/k8s.io/client-go/pkg/apis/apps/types.go:
```diff
@@ -60,6 +60,50 @@ const (
+// StatefulSetUpdateStrategy indicates the strategy that the StatefulSet
+// controller will use to perform updates. It includes any additional parameters
+// necessary to preform the update for the indicated strategy.
+type StatefulSetUpdateStrategy struct {
```
curious does this get auto generated based on previous stuff you added ?

kow3ns replied:
No. You have to add your internal and external types. These are what everything else is generated from.

krmayankk commented on pkg/controller/statefulset/stateful_set_control.go:
```diff
@@ -57,20 +70,196 @@ type defaultStatefulSetControl struct {
+	updateRevision *apps.ControllerRevision,
+	pods []*v1.Pod) (*apps.StatefulSetStatus, error) {
+	// get the current and update revisions of the set. Note that computed
+	currentSet, err := applyRevision(set, currentRevision)
```
consider making applyRevision, method on revision object , so currentRevision.apply(set) ?

kow3ns replied:
I don't think you want to add methods to the API objects.

krmayankk commented on pkg/controller/statefulset/stateful_set_control.go:
```diff
@@ -168,29 +368,91 @@ func (ssc *defaultStatefulSetControl) UpdateStatefulSet(set *apps.StatefulSet, p
+		return &status, nil
+	}
+
+	// we compute the minimum ordinal of the target sequence for a destructive update based on the strategy.
```
what do you mean by destructive update ? burst mode ?

kow3ns replied:
Its destructive in that we have to kill the Pod to recreate it at the correct revision because in place updates are not possible.

krmayankk commented on pkg/controller/statefulset/stateful_set_control.go:
```diff
@@ -168,29 +368,91 @@ func (ssc *defaultStatefulSetControl) UpdateStatefulSet(set *apps.StatefulSet, p
+
+	// we compute the minimum ordinal of the target sequence for a destructive update based on the strategy.
+	updateMin := 0
+	if set.Spec.UpdateStrategy.Type == apps.PartitionStatefulSetStrategyType {
```
can you burst in the partitioned strategy case ? Dont see it handled here

kow3ns replied:
burst is handled in the partitioned case. You can't burst the update, but you can remove ordering from Pod creation. stateful_set_control.go adds a test for this.

krmayankk replied:
I see Partitioned handled only for deletion but not for updation, probably i am not understanding it @kow3ns

kow3ns replied:
The code you point to above decides if a Pod should be deleted. 
1. The method short circuits for OnDeleteStatefulSetStrategy. As in the designm Pods are not terminated with this strategy. Update occurs only when the pod is terminated externally.
1. updateMin is set to 0 by default, we can't have an OnDeleteStatefulSetStrategy here because we've returned control above if did.
1. If the strategy is PartitionedStatefulSetStrategy than the updateMin is set to the Partition, otherwise it's still 0, and we have a RollingUpdateStatefulSetStrategy.
1. In the loop below, we walk backward across the replicas to determine the first one that does not match the update revision. We terminate that replica, and we wait for it to be completely deleted before we attempt to apply another update.
1. Therefore, if we are partitioned, we only considered the update partition sequence for targeted termination, and if we are rolling, we consider the entire sequence. For on delete, we do nothing.
1. Also note that, as per the design, scaling takes precedence and is handled prior to termination for update.

The code below creates new replicas
```golang
replicas[i] = newVersionedStatefulSetPod(currentSet,updateSet,currentRevision.Name,updateRevision.Nam,i)
```

If you look at the internals of that method, you will see that it decides which version a replica should be at based on whether it is being tracked as a current or updated replica and on the selected strategy.

```golang
// newVersionedStatefulSetPod creates a new Pod for a StatefulSet. currentSet is the representation of the set at the
// current revision. updateSet is the representation of the set at the updateRevision. currentRevision is the name of
// the current revision. updateRevision is the name of the update revision. ordinal is the ordinal of the Pod. If the
// returned error is nil, the returned Pod is valid.
func newVersionedStatefulSetPod(currentSet *apps.StatefulSet, updateSet *apps.StatefulSet, currentRevision string, updateRevision string, ordinal int) *v1.Pod {
	if (currentSet.Spec.UpdateStrategy.Type == apps.RollingUpdateStatefulSetStrategyType &&
		ordinal < int(currentSet.Status.CurrentReplicas)) ||
		(currentSet.Spec.UpdateStrategy.Type == apps.PartitionStatefulSetStrategyType &&
			ordinal < int(currentSet.Spec.UpdateStrategy.Partition.Ordinal)) {
		pod := newStatefulSetPod(currentSet, ordinal)
		setPodRevision(pod, currentRevision)
		return pod
	}
	pod := newStatefulSetPod(updateSet, ordinal)
	setPodRevision(pod, updateRevision)
	return pod
}
```
1. If its a current replica we create the replica from the current revision, respecting the partition for the PartitionedStatefulSetStrategy.
1. Otherwise we create it from the update version.

krmayankk replied:
@kow3ns thanks very helpful.

smarterclayton commented on pkg/apis/apps/types.go:
```diff
@@ -60,6 +60,50 @@ const (
 
+// StatefulSetUpdateStrategy indicates the strategy that the StatefulSet
+// controller will use to perform updates. It includes any additional parameters
+// necessary to preform the update for the indicated strategy.
```
typo

smarterclayton commented on pkg/apis/apps/v1beta1/defaults.go:
```diff
@@ -45,6 +48,10 @@ func SetDefaults_StatefulSet(obj *StatefulSet) {
 	}
+	if obj.Spec.RevisionHistoryLimit == nil {
+		obj.Spec.RevisionHistoryLimit = new(int32)
+		*obj.Spec.RevisionHistoryLimit = 2
```
As per @janetkuo's change to daemonsets, let's make this 10.

krmayankk replied:
10 seems like a lot , how did we come up with that number ?

smarterclayton commented on pkg/controller/statefulset/stateful_set_control.go:
```diff
@@ -57,20 +70,196 @@ type defaultStatefulSetControl struct {
+	if err != nil {
+		return err
+	}
+	err = ssc.truncateHistory(set, pods, revisions, currentRevision, updateRevision)
```
daemon set was truncating after update was complete as per @kargakis

smarterclayton replied:
If that's not appropriate here we should comment to that effect

smarterclayton commented on pkg/controller/statefulset/stateful_set_control.go:
```diff
@@ -132,34 +308,58 @@ func (ssc *defaultStatefulSetControl) UpdateStatefulSet(set *apps.StatefulSet, p
-			glog.V(2).Infof("StatefulSet %s is waiting for Pod %s to Terminate", set.Name, replicas[i].Name)
-			return nil
+			glog.V(2).Infof(
+				"StatefulSet %s is waiting for Pod %s to Terminate",
```
All debug in this controller should be at v(4) or above.  Only truly exceptional cases should be below v(4).  Errors should be utilruntime.HandleError()

smarterclayton commented on pkg/controller/statefulset/stateful_set_control.go:
```diff
@@ -168,29 +368,91 @@ func (ssc *defaultStatefulSetControl) UpdateStatefulSet(set *apps.StatefulSet, p
+		status.UpdatedReplicas = 0
+	}
+
+	if set.Status.ObservedGeneration == nil ||
```
Invert this condition to drop a level of indentation.  Also, these are unreadable - you may just want to split the tests into small methods.

janetkuo commented on pkg/apis/apps/types.go:
```diff
@@ -119,6 +174,21 @@ type StatefulSetStatus struct {
+	CurrentRevision string
+
+	// updateRevision is the revision the StatefulSet is attempting to roll out
+	UpdateRevision string
```
I find the comments in proposal very clear and useful, copy them here? https://github.com/kubernetes/community/blob/master/contributors/design-proposals/statefulset-update.md#api-objects

kow3ns commented:
@k8s-bot pull-kubernetes-unit test this

krmayankk commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -255,6 +281,124 @@ func newStatefulSetPod(set *apps.StatefulSet, ordinal int) *v1.Pod {
+// getPatch returns a strategic merge patch that can be applied to restore a StatefulSet to a
+// previous version. If the returned error is nil the patch is valid. The current state that we save is just the
+// PodSpecTemplate. We can modify this later to encompass more state (or less) and remain compatible with previously
+// recorded patches.
```
may be i am missing something, but how can you be compatible with previously recorded patches. If previous revision recorded only podtemplate and later revisions record with pod and volume template,i dont see how we will be compatible ?

kow3ns replied:
If a patch semantically means that you will roll back only the template at the time the revision is recorded it can still be applied to a statefulset where volume claims are also recorded. Rolling back will apply the prior revision without breaking versioning.

krmayankk commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -255,6 +281,124 @@ func newStatefulSetPod(set *apps.StatefulSet, ordinal int) *v1.Pod {
+// to the updateRevision. status's currentRevision is set to updateRevision and its' updateRevision
+// is set to the empty string. status's currentReplicas is set to updateReplicas and its updateReplicas
+// are set to 0.
+func completeRollingUpdate(set *apps.StatefulSet, status *apps.StatefulSetStatus) {
```
what about partitioned rolling update ?

kow3ns replied:
It doesn't complete as per the design document.

kargakis replied:
@krmayankk you should be able to estimate if a partial rollout is complete with 
```
foo.spec.replicas-foo.spec.updateStrategy.partition.ordinal == foo.status.updatedReplicas &&
foo.spec.replicas == foo.status.readyReplicas
```
There is definitely value in communicating that via a Condition in the future.

krmayankk commented on pkg/controller/statefulset/stateful_set_control.go:
```diff
@@ -36,18 +37,30 @@ type StatefulSetControlInterface interface {
 	// exit exceptionally at any point provided they wish the update to be re-run at a later point in time.
 	UpdateStatefulSet(set *apps.StatefulSet, pods []*v1.Pod) error
+	// ListRevisions returns a array of the ControllerRevisions that represent the revisions of set. If the returned
+	// error is not nil, the returns slice of ControllerRevisions is valid.
```
nit: `error is nil, the returned slice of the ControllerRevisions is valid.`

krmayankk commented on pkg/controller/statefulset/stateful_set_control.go:
```diff
@@ -57,20 +70,203 @@ type defaultStatefulSetControl struct {
+	}
+
+	// maintain the set's revision history limit
+	return ssc.truncateHistory(set, pods, revisions, currentRevision, updateRevision)
```
as i understand the flow, it is sync--->syncStatefulSet--->UpdateStatefulSet.  We are doing adoptOrphanRevisions in sync while truncateHistory inside UpdateStatefulSet.  The split wrt revisions is little odd. We could move adoption inside Update or move the truncate out of UpdateStatefulSet so Update is just about updating and Revision management is outside this function.

kow3ns replied:
The statefulset client isn't in that scope. All of the Pod adoption is done in the outer sync loop. I put the logic for revision adoption there as well. If we want to discuss a refactor than we should move both (and I'm not opposed to that idea), but that should be another PR.

kargakis replied:
This will not be executed at the end of the rollout if that's what you are shooting for here, right?

krmayankk commented on pkg/controller/statefulset/stateful_set_control.go:
```diff
@@ -57,20 +70,203 @@ type defaultStatefulSetControl struct {
+	history = history[:int(*set.Spec.RevisionHistoryLimit)]
+	for i := 0; i < len(history); i++ {
+		if err := ssc.controllerHistory.DeleteControllerRevision(history[i]); err != nil {
+			return err
```
why do we bail out on first error ? We should attempt to delete all and log errors for the revisions we couldnt .

kow3ns replied:
No. There is a backoff throttle in the outer sync loop that causes retries to be delayed. That is why we bail out early on errors. If the API Server is stressed we shouldn't continue to pound it. We should back off and attempt again later.

krmayankk commented on pkg/controller/statefulset/stateful_set_control.go:
```diff
@@ -57,20 +70,203 @@ type defaultStatefulSetControl struct {
+}
+
+// truncateHistory truncates any non-live ControllerRevisions in revisions from set's history. The UpdateRevision and
+// CurrentRevision in set's Status are considered to be live, as are any revisions associated with the Pods in pods.
```
nit: reword to say `Any revision at set's CurrentRevision or set's UpdateRevision are considered live. Also any revision associated with the Pods in `pods` is also live`

krmayankk commented on pkg/controller/statefulset/stateful_set_control.go:
```diff
@@ -57,20 +70,203 @@ type defaultStatefulSetControl struct {
+func (ssc *defaultStatefulSetControl) truncateHistory(
+	set *apps.StatefulSet,
+	pods []*v1.Pod,
+	revisions []*apps.ControllerRevision,
```
is revisions already sorted, please mention in the comment

kow3ns replied:
It is sorted, directly above the invocation site, but I will add a comment to the helper methods declaring the expections about the parameters.

krmayankk commented on pkg/controller/statefulset/stateful_set_control.go:
```diff
@@ -57,20 +70,203 @@ type defaultStatefulSetControl struct {
+	}
+	// delete any non-live history to maintain the revision limit.
+	history = history[:int(*set.Spec.RevisionHistoryLimit)]
+	for i := 0; i < len(history); i++ {
```
is history is not already sorted, we should sort before attempting delete

kow3ns replied:
It is sorted prior to invocation. I added a comment about methods expectations of the supplied parameter.

krmayankk commented on pkg/controller/statefulset/stateful_set_control.go:
```diff
@@ -57,20 +70,203 @@ type defaultStatefulSetControl struct {
+
+// getSetRevisions returns the current and update ControllerRevisions for set. This method may create a new revision,
+// or modify the Revision of an existing revision if an update to set is detected.
+func (ssc *defaultStatefulSetControl) getSetRevisions(
```
call this `getStatefulSetRevisions`, getSetRev is confusing

krmayankk commented on pkg/controller/statefulset/stateful_set_control.go:
```diff
@@ -57,20 +70,203 @@ type defaultStatefulSetControl struct {
+	history.SortControllerRevisions(revisions)
+
+	// create a new revision from the current set
+	updateRevision, err := newRevision(set, nextRevision(revisions))
```
may be add a comment. It seems we will always make a new revision for statefulset during update, even though equivalent previous revisions exist , right ?

kow3ns replied:
It may not create it

krmayankk commented on pkg/controller/statefulset/stateful_set_control.go:
```diff
@@ -57,20 +70,203 @@ type defaultStatefulSetControl struct {
+	equalRevisions := history.FindEqualRevisions(revisions, updateRevision)
+	equalCount := len(equalRevisions)
+
+	if equalCount > 0 && history.EqualRevision(revisions[revisionCount-1], equalRevisions[equalCount-1]) {
```
since you are comparing with last equalRevisions, the equalRevisions should be sorted.

kow3ns replied:
it is sorted by I will add a comment about the methods expectation of the supplied parameter.

krmayankk commented on pkg/controller/statefulset/stateful_set_control.go:
```diff
@@ -57,20 +70,203 @@ type defaultStatefulSetControl struct {
+			return nil, nil, err
+		}
+	} else {
+		//if there is no equivalent revision we create a new one
```
it looks like this is the only case when we should be creating a new revision, move the line `updateRevision, err := newRevision(set, nextRevision(revisions))` here

kow3ns replied:
We need to create the revision in order to search for any equal revisions prior to the first conditional branch test.

krmayankk commented on pkg/controller/statefulset/stateful_set_control.go:
```diff
@@ -130,36 +301,60 @@ func (ssc *defaultStatefulSetControl) UpdateStatefulSet(set *apps.StatefulSet, p
+			glog.V(4).Infof("StatefulSet %s is recreating failed Pod %s", set.Name, replicas[i].Name)
 			if err := ssc.podControl.DeleteStatefulPod(set, replicas[i]); err != nil {
-				return err
+				return &status, err
```
confused why we are not doing the monotonic check here

kow3ns replied:
We need to delete any failed Pods. This doesn't count as a termination. You should look at the deletion handler for Pods in the api server to understand it fully. When you delete a failed Pod you will just remove it from storage. Its already terminated. All our other deletes will set the termination grace and we have to wait for the Pod to turn down.

smarterclayton commented on pkg/controller/statefulset/stateful_set_status_updater.go:
```diff
@@ -0,0 +1,78 @@
+func (ssu *realStatefulSetStatusUpdater) UpdateStatefulSetStatus(
+	set *apps.StatefulSet,
+	status *apps.StatefulSetStatus) error {
+	return retry.RetryOnConflict(retry.DefaultBackoff, func() error {
```
DefaultRetry - backoff isn't needed for conflict updates since it's usually a fixed number of clients trying to react

smarterclayton commented:
Some follow up e2es (can be a subsequent PR):

1. Update, rollback, update again (preserve invariants)
2. Partition including boundary condition
3. Pod replacement during update

falenn commented:
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  curtis.l.bates@gmail.com
    Domain imwiz.com has exceeded the max emails per hour (151/150 (100%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
------ The body of the message is 13209 characters long; only the first
------ 5000 or so are included here.
Received: from o6.sgmail.github.com ([192.254.113.101]:2131)
	by box969.bluehost.com with esmtps (TLSv1.2:ECDHE-RSA-AES128-GCM-SHA256:128)
	(Exim 4.87)
	(envelope-from <bounces+848413-5c7e-dev=imwiz.com@sgmail.github.com>)
	id 1dGU4O-000OQ0-3Y
	for dev@imwiz.com; Thu, 01 Jun 2017 11:39:37 -0600
DKIM-Signature: v=1; a=rsa-sha1; c=relaxed/relaxed; d=github.com; 
	h=from:reply-to:to:cc:in-reply-to:references:subject:mime-version:content-type:content-transfer-encoding:list-id:list-archive:list-post:list-unsubscribe; 
	s=s20150108; bh=Ssm8bhe0qVDVvQl3OaRSn8Zozt8=; b=HuFfWRqRuFzKdPU/
	dYf6YulvlZxmG57g4ebO4twjQ+PWYT4CwDZJIAFHuwjslThcuGu5mRc+En+A16qP
	LwAnOlfTU/kL88UQVZfXLctsbW9Wgu75w+zVsjbSMOrkmpG6ciQblcCbZtyKBr3J
	dF5WKraU9ylCUV9vWEHpMLcos+o=
Received: by filter1012p1mdw1.sendgrid.net with SMTP id filter1012p1mdw1-9762-59305145-79
        2017-06-01 17:39:17.722994155 +0000 UTC
Received: from github-smtp2a-ext-cp1-prd.iad.github.net (github-smtp2a-ext-cp1-prd.iad.github.net [192.30.253.16])
	by ismtpd0001p1iad1.sendgrid.net (SG) with ESMTP id ynyQxFQxRMOqFJUj6-e1zQ
	for <dev@imwiz.com>; Thu, 01 Jun 2017 17:39:17.677 +0000 (UTC)
Date: Thu, 01 Jun 2017 10:39:17 -0700
From: k8s-ci-robot <notifications@github.com>
Reply-To: kubernetes/kubernetes <reply@reply.github.com>
To: kubernetes/kubernetes <kubernetes@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Message-ID: <kubernetes/kubernetes/pull/46669/c305566275@github.com>
In-Reply-To: <kubernetes/kubernetes/pull/46669@github.com>
References: <kubernetes/kubernetes/pull/46669@github.com>
Subject: Re: [kubernetes/kubernetes] implements StatefulSet update (#46669)
Mime-Version: 1.0
Content-Type: multipart/alternative;
 boundary="--==_mimepart_593051453ff56_516e3fb820ee9c345387c";
 charset=UTF-8
Content-Transfer-Encoding: 7bit
Precedence: list
X-GitHub-Sender: k8s-ci-robot
X-GitHub-Recipient: falenn
X-GitHub-Reason: subscribed
List-ID: kubernetes/kubernetes <kubernetes.kubernetes.github.com>
List-Archive: https://github.com/kubernetes/kubernetes
List-Post: <mailto:reply@reply.github.com>
List-Unsubscribe: <mailto:unsub+000ab60a1c9180f34395f3fd661c433d671371f13c8bdb4c92cf000000011548134592a169ce0ddab907@reply.github.com>,
 <https://github.com/notifications/unsubscribe/AAq2Csp_hIdh9_WcPC8bFHyy08QTxOBFks5r_vdFgaJpZM4NrFxI>
X-Auto-Response-Suppress: All
X-GitHub-Recipient-Address: dev@imwiz.com
X-SG-EID: APO41b8ovafPb3SK9rw3vGS2Kq45kgpYx4y17m0ryg0AdzL5XlG6oweTR9hAZElS7IV8zaHpX/eJhp
 woHUS9gPBNfvEzykeVA6W6BWWc7oF0V0yyp+Hdidf7pVRekepuITytQZWSrsnEyiEyZQomvOGO0Nwy
 T+eKmHTPFfqCUzKubA+S2cofBUQXSBCi+miCZNSquWZJBntk16byzCCWtQ==
X-Spam-Status: No, score=-3.1
X-Spam-Score: -30
X-Spam-Bar: ---
X-Ham-Report: Spam detection software, running on the system "box969.bluehost.com",
 has NOT identified this incoming email as spam.  The original
 message has been attached to this so you can view it or label
 similar future email.  If you have any questions, see
 root\@localhost for details.
 
 Content preview:  @kow3ns: The following test(s) **failed**: Test name | Commit
    | Details | Rerun command --- | --- | --- | --- pull-kubernetes-federation-e2e-gce
    | 069f7e7b9f930f01e7ebad2a2f55618105301b85 | [link](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46669/pull-kubernetes-federation-e2e-gce/7870/)
    | `@k8s-bot pull-kubernetes-federation-e2e-gce test this` pull-kubernetes-bazel
    | 3b713de6bfaf2d2876a14d315e18713b1a000ecd | [link](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46669/pull-kubernetes-bazel/24296/)
    | `@k8s-bot pull-kubernetes-bazel test this` pull-kubernetes-node-e2e | 3b713de6bfaf2d2876a14d315e18713b1a000ecd
    | [link](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46669/pull-kubernetes-node-e2e/33303/)
    | `@k8s-bot pull-kubernetes-node-e2e test this` pull-kubernetes-kubemark-e2e-gce
    | 3b713de6bfaf2d2876a14d315e18713b1a000ecd | [link](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46669/pull-kubernetes-kubemark-e2e-gce/33164/)
    | `@k8s-bot pull-kubernetes-kubemark-e2e-gce test this` pull-kubernetes-e2e-kops-aws
    | 3b713de6bfaf2d2876a14d315e18713b1a000ecd | [link](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46669/pull-kubernetes-e2e-kops-aws/28417/)
    | `@k8s-bot pull-kubernetes-e2e-kops-aws test this` pull-kubernetes-unit
   | 3b713de6bfaf2d2876a14d315e18713b1a000ecd | [link](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46669/pull-kubernetes-unit/34157/)
    | `@k8s-bot pull-kubernetes-unit test this` pull-kubernetes-verify | 3b713de6bfaf2d2876a14d315e18713b1a000ecd
    | [link](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46669/pull-kubernetes-verify/33170/)
    | `@k8s-bot pull-kubernetes-verify test this` pull-kubernetes-e2e-gce-etcd3
    | 3b713de6bfaf2d2876a14d315e18713b1a000ecd | [link](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46669/pull-kubernetes-e2e-gce-etcd3/33638/)
    | `@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this` [...] 
 
 Content analysis details:   (-3.1 points, 4.0 required)
 
  pts rule name              description
 ---- ---------------------- --------------------------------------------------
 -0.0 RP_MATCHES_RCVD        Envelope sender domain matches handover relay domain
 -2.8 RCVD_IN_MSPIKE_H2      RBL: Average reputation (+2)
                             [192.254.113.101 listed in wl.mailspike.net]
 -0.0 SPF_PASS               SPF: sender matches SPF record
  0.0 HTML_MESSAGE           BODY: HTML included in message
 -0.1 DKIM_VALID_AU          Message has a valid DKIM or DK signature from author's
                             domain
 -0.1 DKIM_VALID             Message has at least one valid DKIM or DK signature
  0.1 DKIM_SIGNED            Message has a DKIM or DK signature, not necessarily valid
 -0.2 AWL                    AWL: Adjusted score from AWL reputation of From: address
X-Spam-Flag: NO

----==_mimepart_593051453ff56_516e3fb820ee9c345387c
Content-Type: text/plain;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

@kow3ns: The following test(s) **failed**:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
pull-kubernetes-federation-e2e-gce | 069f7e7b9f930f01e7ebad2a2f55618105301b85 | [link](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46669/pull-kubernetes-federation-e2e-gce/7870/) | `@k8s-bot pull-kubernetes-federation-e2e-gce test this`
pull-kubernetes-bazel | 3b713de6bfaf2d2876a14d315e18713b1a000ecd | [link](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46669/pull-kubernetes-bazel/24296/) | `@k8s-bot pull-kubernetes-bazel test this`
pull-kubernetes-node-e2e | 3b713de6bfaf2d2876a14d315e18713b1a000ecd | [link](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46669/pull-kubernetes-node-e2e/33303/) | `@k8s-bot pull-kubernetes-node-e2e test this`
pull-kubernetes-kubemark-e2e-gce | 3b713de6bfaf2d2876a14d315e18713b1a000ecd | [link](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46669/pull-kubernetes-kubemark-e2e-gce/33164/) | `@k8s-bot pull-kubernetes-kubemark-e2e-gce test this`
pull-kubernetes-e2e-kops-aws | 3b713de6bfaf2d2876a14d315e18713b1a000ecd | [link](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46669/pull-kubernetes-e2e-kops-aws/28417/) | `@k8s-bot pull-kubernetes-e2e-kops-aws test this`
pull-kubernetes-unit | 3b713de6bfaf2d2876a14d315e18713b1a000ecd | [link](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46669/pull-kubernetes-unit/34157/) | `@k8s-bot pull-kubernetes-unit test this`
pull-kubernetes-verify | 3b713de6bfaf2d2876a14d315e18713b1a000ecd | [link](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46669/pull-kubernetes-verify/33170/) | `@k8s-bot pull-kubernetes-verify test this`
pull-kubernetes-e2e-gce-etcd3 | 3b713de6bfaf2d2876a14d315e18713b1a000ecd | [link](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46669/pull-kubernetes-e2e-gce-etcd3/33638/) | `@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this`

[Full PR test history](https://k8s-gubernator.appspot.com/pr/46669). [Your PR dashboard](https://k8s-gubernator.appspot.com/pr/kow3ns). Please help us cut down on flakes by [linking to](https://github.com/kubernetes/community/blob/master/contributors/devel/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes/kubernetes/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-request-commands.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

-- 
You are receiving this because you are subscribed to this thread.
Reply to this email directly or view it on GitHub:
https://github.com/kubernetes/kubernetes/pull/46669#issuecomment-305566275
----==_mimepart_593051453ff56_516e3fb820ee9c345387c
Content-Type: text/html;
 charset=UTF-8
Content-Transfer-Encoding: quoted-printable

<p><a href=3D"https://github.com/kow3ns" class=3D"user-mention">@kow3ns</a>=
: The following test(s) <strong>failed</strong>:</p>
<table>
<thead>
<tr>
<th>Test name</th>
<th>Commit</th>
<th>Details</th>
<th>Rerun command</th>
</tr>
</thead>
<tbody>
<tr>
<td>pull-kubernetes-federation-e2e-gce</td>
<td><a href=3D"https://github.com/kubernetes/kubernetes/commit/069f7e7b9f93=
0f01e7ebad2a2f55618105301b85" class=3D"commit-link"><tt>069f7e7</tt></a></t=
d>
<td><a href=3D"https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/=
pr-logs/pull/46669/pull-kubernetes-federation-e2e-gce/7870/">link</a></td>
<td><code>@k8s-bot pull-kubernetes-federation-e2e-gce test this</code></td>
</tr>
<tr>
<td>pull-kubernetes-bazel</td>
<td><a href=3D"https://github.com/kubernetes/kubernetes/commit/3b713de6bfaf=
2d2876a14d315e18713b1a000ecd" class=3D"commit-link"><tt>3b713de</tt></a></t=
d>
<td><a href=3D"https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/=
pr-logs/pull/46669/pull-kubernetes-bazel/24296/">link</a></td>
<td><code>@k8s-bot pull-kubernetes-bazel test this</code></td>
</tr>
<tr>
<td>pull-kubernetes-node-e2e</td>
<td><a href=3D"https://github.com/kubernetes/kubernetes/commit/3b713de6bfaf=
2d2876a14d315e18713b1a000ecd" class=3D"commit-link"><tt>3b713de</tt></a></t=
d>
<td><a href=3D"https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/=
pr-logs/pull/46669/pull-kubernetes-node-e2e/33303/">link</a></td>
<td><code>@k8s-bot pull-kubernetes-node-e2e test this</code></td>
</tr>
<tr>
<td>pull-kubernete

smarterclayton commented on pkg/controller/statefulset/stateful_set_status_updater.go:
```diff
@@ -53,7 +54,8 @@ type realStatefulSetStatusUpdater struct {
 	status *apps.StatefulSetStatus) error {
-	return retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+	// don't wait due to limited number of clients, but backoff after the default number of steps
+	return retry.RetryOnConflict(wait.Backoff{Steps: retry.DefaultBackoff.Steps}, func() error {
```
There should be an explicit DefaultRetry var you can use.

smarterclayton commented:
/approve

This is pretty close - I'd like to get it in and the e2e tests in as soon as possible.

smarterclayton commented:
The full e2e tests don't have to be in the same PR.

smarterclayton commented:
Basically LGTM. Our fallback path for 1.7 is if necessary to disable the strategy change if we decide we're not ready to release it.

krmayankk commented:
@kow3ns @smarterclayton i would be happy to write a few e2e which are lower priority plus the additional refactoring i was referring to earlier in code. I will review once more today evening, although i still have some unaswered comments. Mainly i dont understand how partitioning is working with current changes.

kow3ns commented:
@k8s-bot pull-kubernetes-kubemark-e2e-gce test this

kow3ns commented on pkg/apis/apps/validation/validation.go:
```diff
@@ -75,6 +75,40 @@ func ValidateStatefulSetSpec(spec *apps.StatefulSetSpec, fldPath *field.Path) fi
+	switch spec.UpdateStrategy.Type {
+	case "":
+		allErrs = append(allErrs, field.Required(fldPath.Child("updateStrategy"), ""))
+	case apps.OnDeleteStatefulSetStrategyType, apps.RollingUpdateStatefulSetStrategyType:
```
I turned them off because it might be confusing. I'm open to suggestion about the best practice wrt to leaving the su- field allowable. I gave it some thought previously and thought that ignoring it might be the best course of action to allow easy type changes. After further thought it seems like it might be confusing. A user could get a SatefulSet and see a partition for rolling update and wonder why it isn't respected. During the design phase we kicked the idea around of applying a partition directly in the spec and we landed on using the structs. I'm leaning in favor of failing validation if the a partition is applied to a strategy that can not support it. It does not represent a valid API object based on the design.

kargakis replied:
@kow3ns does your comment hold true? I see you are validating now.

kow3ns replied:
Yes I added the validation after @smarterclayton mentioned it and I thought it over. If it does become painful we can remove the validation without breaking the interface. We can't add it later without doing an API bump.

kow3ns commented:
@krmayankk I think all of the review comments you have posted thus far have been either answered or implemented. I added greater detail about how partitioning works.

kow3ns commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

kow3ns commented:
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

kow3ns commented:
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

kow3ns commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

kow3ns commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

fejta commented:
federation failure ref: #46827

krmayankk commented on pkg/apis/apps/validation/validation.go:
```diff
@@ -75,6 +75,48 @@ func ValidateStatefulSetSpec(spec *apps.StatefulSetSpec, fldPath *field.Path) fi
+				allErrs, field.Invalid(
+					fldPath.Child("updateStrategy").Child("partition").Child("ordinal"),
+					spec.UpdateStrategy.Partition.Ordinal,
+					"must be less than 'replicas'"))
```
the check above suggests the partition can be equal to replicas but the error says otherwise. We need to fix the check above

krmayankk commented on pkg/apis/apps/validation/validation_test.go:
```diff
@@ -202,6 +246,56 @@ func TestValidateStatefulSet(t *testing.T) {
+					Type: apps.PartitionStatefulSetStrategyType,
+					Partition: func() *apps.PartitionStatefulSetStrategy {
+						return &apps.PartitionStatefulSetStrategy{Ordinal: 4}
+					}()},
```
add a test for specifying partition in non-partitioned strategy ?

kow3ns replied:
Already have that test
```golang
"partitioned rolling update": {
			ObjectMeta: metav1.ObjectMeta{Name: "abc-123", Namespace: metav1.NamespaceDefault},
			Spec: apps.StatefulSetSpec{
				PodManagementPolicy: apps.OrderedReadyPodManagement,
				Selector:            &metav1.LabelSelector{MatchLabels: validLabels},
				Template:            validPodTemplate.Template,
				Replicas:            3,
				UpdateStrategy: apps.StatefulSetUpdateStrategy{Type: apps.RollingUpdateStatefulSetStrategyType,
					Partition: func() *apps.PartitionStatefulSetStrategy {
						return &apps.PartitionStatefulSetStrategy{Ordinal: 2}
					}()},
			},
		},
```

krmayankk commented on pkg/controller/history/controller_history.go:
```diff
@@ -0,0 +1,487 @@
+}
+
+// NewControllerRevision returns the a ControllerRevision with a ControllerRef pointing parent and indicating that
+// parent is of parentKing. The ControllerRevision has labels matching selector, contains Data equal to data, and
```
nit: parentKind

kargakis commented on pkg/apis/apps/validation/validation.go:
```diff
@@ -75,6 +75,48 @@ func ValidateStatefulSetSpec(spec *apps.StatefulSetSpec, fldPath *field.Path) fi
+					fmt.Sprintf("only allowed for updateStrategy '%s'", apps.PartitionStatefulSetStrategyType)))
+		}
+	case apps.PartitionStatefulSetStrategyType:
+		if spec.UpdateStrategy.Partition == nil {
```
I expect the approach of using a different strategy for partitioning to be more confusing and painful as opposed to this being an option for rollingUpdate. If I want to build an operator on top of StatefulSets, I am going to lose a lot of UX with the current approach. Assuming we add Conditions in the future and surface one once an update is complete, it is going to be much easier with a single strategy. I am thinking of post hooks. Now, you can't have a post-hook and re-use the StatefulSet controller to set a complete Condition just by switching Partition from 0 (which means all Pods have been updated) to nil. You have to make the operator set the Condition.

cc: tnozicka

kargakis replied:
xref: https://github.com/kubernetes/community/pull/503#issuecomment-305850408

kow3ns replied:
I can get on board with the approach of adding  the partition to rolling update, but I think we can send the condition from the  controller in either case. In order to perform a staged rollout you can use patches like the one below while decrementing the ordinal.
```json
{
	"updateStrategy": {
		"type": "Partitioned",
		"ordinal": <ordinal>
	}
}
```

Since the partitioned strategy never completes, no signal will be sent, via a Condition or otherwise, until user switches the strategy back to rolling update or on delete. So the last applied patch will look like this.

```yaml
{
	"updateStrategy": {
		"type": "RollingUpdate"
	}
}
```

At this point the controller will complete the update and send any signals (including Conditions).

What I like about this approach is that operators, human or otherwise, have an opportunity to validate the deployed product prior to triggering the controller to signal completion. Readiness checks are good for determining if the product is running, but this would allow for adding a smoke test prior completing the update.

Again, I don't feel super strongly about it

kargakis replied:
> What I like about this approach is that operators, human or otherwise, have an opportunity to validate the deployed product prior to triggering the controller to signal completion. Readiness checks are good for determining if the product is running, but this would allow for adding a smoke test prior completing the update.

Not sure why this is different in the case partitioning is an option of the rolling strategy. A smoke test prior to completing the update, or sending out an e-mail, or any kind of post-hook action could be possible to trigger by setting the ordinal to the size of the statefulset (sts.spec.replicas). At that point, the controller has moved all replicas to use the updated revision but it still waits for a higher-level orchestrator to set the ordinal to nil in order for the update to be considered as complete.

Another reason for moving this as an option vs a separate strategy is that in the future adding new strategies will make validation and UX even more complex and you will likely need to create an additional partition strategy for the new strategy. Take for example Deployments, we support Rolling and Recreate. Adding a single Partitioned strategy is not going to work for both, adding a PartitionedRolling and PartitionedRecreate will be a UX hit.

kargakis replied:
> by setting the ordinal to the size of the statefulset (sts.spec.replicas)

Sorry, this was meant to be zero.

krmayankk replied:
Would we have staged rollout marked as completed when we move partition inside the RollingUpdate struct ? I would like to see the same method of determining whether a Update completed or not , whether we are using staged or not. I would canary for 2 days or even 1 week, and having to determine the update completed based on looking at different fields is confusing not to say that a external tooling/portal showing the staged rollout as incomplete until the partitioning is removed and normal rolling update strategy is applied.

kow3ns replied:
@krmayankk No. That is not what @kargakis is suggesting. We would have a RollingUpdate that completes when all Pods are at a consistent user declared state indicated by both (in this most recent commit) the updateRevision and the currentRevision. We would remove the PartitionStrategy altogether.

kow3ns replied:
@kargakis, @smarterclayton  Do you feel strongly enough about this to gate LGTM? I'd be fine discussing this further offline and issuing PRs for removal in both the design and implementation after the fact, if that's where we land. If you think its a show stopper, I'd also be fine removing it in this PR provided we have general agreement. We can issue a PR to update the design concurrently.

kargakis replied:
I don't want to block this PR on this and I am fine with follow-ups if we decide to change.

smarterclayton replied:
I'd like to merge and then iterate.  I agree with the points @kargakis is making but once this lands we can do an offline discussion and reach closure in time for release cut.

kargakis replied:
@kow3ns please open an 1.7 issue so we can track

kargakis commented on pkg/apis/apps/validation/validation.go:
```diff
@@ -75,6 +75,48 @@ func ValidateStatefulSetSpec(spec *apps.StatefulSetSpec, fldPath *field.Path) fi
+				field.Required(
+					fldPath.Child("updateStrategy").Child("partition"),
+					fmt.Sprintf("required for updateStrategy '%s'", apps.PartitionStatefulSetStrategyType)))
+			break
```
redundant statement

kargakis replied:
disregard

kargakis commented on pkg/apis/apps/validation/validation.go:
```diff
@@ -75,6 +75,48 @@ func ValidateStatefulSetSpec(spec *apps.StatefulSetSpec, fldPath *field.Path) fi
+			apivalidation.ValidateNonnegativeField(
+				int64(spec.UpdateStrategy.Partition.Ordinal),
+				fldPath.Child("updateStrategy").Child("partition").Child("ordinal"))...)
+		if spec.UpdateStrategy.Partition.Ordinal > spec.Replicas {
```
This means that an autoscaler may break on a StatefulSet that is canarying. I would suggest looser validation for this one.

kow3ns replied:
SGTM. The semantics are achieved as long as it is non-negative. Greater than can be the same as equal to.

kargakis commented on pkg/apis/apps/validation/validation.go:
```diff
@@ -113,20 +155,21 @@ func ValidateStatefulSet(statefulSet *apps.StatefulSet) field.ErrorList {
 
 	if !reflect.DeepEqual(statefulSet.Spec, oldStatefulSet.Spec) {
-		allErrs = append(allErrs, field.Forbidden(field.NewPath("spec"), "updates to statefulset spec for fields other than 'replicas' and 'containers' are forbidden."))
+		allErrs = append(allErrs, field.Forbidden(field.NewPath("spec"), "updates to statefulset spec for fields other than 'replicas', 'template', and 'updateStrategy' are forbidden."))
```
Is it only the volumeClaimTemplate that is not valid on updates now?

kow3ns replied:
So we made the PodTemplateSpec mutable, but nothing else. This means that labels and annotations still can't be added. I'm erring on the side of caution with respect to opening up mutation, but have been considering your previous comments that StatefulSet should be as mutable as DaemonSet an ReplicaSet. My thought is that we should reduce constraints over time rather than all at once  because once we make something mutable making it immutable again is a API breaking change.

krmayankk replied:
what about serviceName ?

kow3ns replied:
Its mutable. You should look at the implementation of the main control loop. If you mutate it we will update the pod to conform in this code.
```golang
// updateIdentity updates pod's name, hostname, and subdomain to conform to set's name and headless service.
func updateIdentity(set *apps.StatefulSet, pod *v1.Pod) {
	pod.Name = getPodName(set, getOrdinal(pod))
	pod.Namespace = set.Namespace
	pod.Spec.Hostname = pod.Name
	pod.Spec.Subdomain = set.Spec.ServiceName
}
```
If you are consuming it via the downward API you will see the change. You can get yourself into trouble playing with it, but you can get yourself into even more trouble playing with the Selector of DaemonSets, ReplicaSets, and Deployments, and, as there are, at least for Deployments and ReplicaSets, valid use cases for that mutation we allow it.

kargakis commented on pkg/controller/history/controller_history.go:
```diff
@@ -0,0 +1,487 @@
+	"k8s.io/kubernetes/pkg/client/retry"
+)
+
+// HashLabel is the label used to indicate the hash value of a ControllerRevision's Data.
```
godoc is not matching the const

kargakis commented on pkg/controller/history/controller_history.go:
```diff
@@ -0,0 +1,487 @@
+// HashLabel is the label used to indicate the hash value of a ControllerRevision's Data.
+const ControllerRevisionHashLabel = "controller.kubernetes.io/hash"
+
+// ControllerRevisionName returns the Name for a ControllerRevision in the form prefix-hash-revision. If the length
```
prefix-hash instead of prefix-hash-revision?

kargakis commented on pkg/controller/history/controller_history.go:
```diff
@@ -0,0 +1,487 @@
+			return nil
+		}
+		clone.Revision = newRevision
+		_, updateErr := rh.client.Apps().ControllerRevisions(clone.Namespace).Update(clone)
```
Assign the updated copy in clone.

kargakis commented on pkg/controller/history/controller_history.go:
```diff
@@ -0,0 +1,487 @@
+	// Use strategic merge patch to add an owner reference indicating a controller ref
+	return rh.client.Apps().ControllerRevisions(parent.GetNamespace()).Patch(revision.GetName(),
+		types.StrategicMergePatchType, []byte(fmt.Sprintf(
+			`{"metadata":{"ownerReferences":[{"apiVersion":"%s","kind":"%s","name":"%s","uid":"%s","controller":true,"blockOwnerDeletion":true}],"uid":"%s"}}`,
```
You also need to set a finalizer, see https://github.com/kubernetes/kubernetes/pull/45764

kow3ns replied:
So I'm setting the finalizer during construction and I've added it the path as well, but I'm using a StrategicMergePatch so shouldn't I be able to get away with only setting it on construction? I mean the implementation I'm using works so I'll leave it be, but it seems like we shouldn't have to reset  the finalizers in this case or in the case of Deployments Orphaning ReplicaSets.

kargakis replied:
Yeah, this is needed in ReplicaSets so the deletion can cascade to Pods but not in ControllerRevisions (at least for now) - disregard.

kargakis commented on pkg/controller/history/controller_history.go:
```diff
@@ -0,0 +1,487 @@
+	return released, err
+}
+
+type fakeHistory struct {
```
Since this is implementing an interface, can't it live side-by-side in testing code?

kow3ns replied:
I export it into StatefulSet. It seems like most fakes in the source code do not live in testing files. Also, we may want to add methods to aid in testing and export it as a class later. All the code that uses the interface now will still work if we change it to a class that implements the interface.

kargakis commented on pkg/controller/statefulset/stateful_set_control.go:
```diff
@@ -17,12 +17,13 @@ limitations under the License.
 	"k8s.io/kubernetes/pkg/api/v1"
 	apps "k8s.io/kubernetes/pkg/apis/apps/v1beta1"
+	"k8s.io/kubernetes/pkg/controller/history"
+	"math"
```
Move stdlib imports back on the top

kargakis commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -255,6 +281,124 @@ func newStatefulSetPod(set *apps.StatefulSet, ordinal int) *v1.Pod {
+// Generation or if any of the status's fields do not match those of set's status.
+func inconsistentStatus(set *apps.StatefulSet, status *apps.StatefulSetStatus) bool {
+	return set.Status.ObservedGeneration == nil ||
+		*status.ObservedGeneration > *set.Status.ObservedGeneration ||
```
set.Generation > *status.ObservedGeneration seems to be missing

kow3ns replied:
The Generation of the set is stored in status prior to invocation. That is, status carries the Generation of the StatefulSet as its ObservedGeneration.

kargakis replied:
Still, I think it's better to add it here, too. It makes the helper not depend on the generation being set prior to its invocation.

kargakis commented on pkg/controller/statefulset/stateful_set_control.go:
```diff
@@ -57,20 +70,222 @@ type defaultStatefulSetControl struct {
+	currentRevision *apps.ControllerRevision,
+	updateRevision *apps.ControllerRevision,
+	pods []*v1.Pod) (*apps.StatefulSetStatus, error) {
+	// get the current and update revisions of the set. Note that computed
```
unfinished comment?

kargakis commented on pkg/apis/apps/v1beta1/defaults.go:
```diff
@@ -30,6 +30,10 @@ func SetDefaults_StatefulSet(obj *StatefulSet) {
 	}
+
+	if obj.Spec.UpdateStrategy.Type == "" {
+		obj.Spec.UpdateStrategy.Type = RollingUpdateStatefulSetStrategyType
```
The whole point of having "OnDelete" as a strategy is to be backwards-compatible. This is a breaking change.

cc: @kubernetes/sig-apps-api-reviews

kow3ns replied:
I'm fine with defaulting to OnDelete. The plan would be to maintain this for the lifetime of v1beta1 an then move it to RollingUpdate in GA (or whatever the next API version we support is)?

kargakis replied:
Yes, we can default to RollingUpdate on the next API version, whatever that is.

kow3ns replied:
SGTM

kow3ns commented:
@kargakis the finalizer did not appear to work. It doesn't even pass unit tests so I pulled it.

janetkuo commented on pkg/controller/history/controller_history.go:
```diff
@@ -0,0 +1,490 @@
+			return nil
+		}
+		clone.Revision = newRevision
+		updated, updateErr := rh.client.Apps().ControllerRevisions(clone.Namespace).Update(clone)
```
`Apps()` is deprecated, `AppsV1beta1` should be used instead

kow3ns replied:
I am planning on updating statefulset and history to use AppsV1beta1 in a follow up PR. I think it is best to update them all, simultaneously, with a PR to do explicitly that, but I'd be willing to update history now if there is a strong reason.

janetkuo replied:
It can be done in a follow up PR.

kow3ns replied:
I opened #46922, and I think we should do this prior to 1.7 release.

kow3ns commented:
@k8s-bot pull-kubernetes-unit test this

kargakis commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -255,6 +281,124 @@ func newStatefulSetPod(set *apps.StatefulSet, ordinal int) *v1.Pod {
+		status.CurrentReplicas = status.UpdatedReplicas
+		status.CurrentRevision = status.UpdateRevision
+		status.UpdateRevision = ""
+		status.UpdatedReplicas = 0
```
xref: https://github.com/kubernetes/community/pull/503#issuecomment-305832012

There is also precedent with the field in Deployments and DaemonSets.

kow3ns replied:
It seems counter intuitive to me that when an update completes successfully that the reported current version of the deployed product would be inconsistent with the version newly deployed product. 

Consider the following scenario.

As implemented, when a user deploys a StatefulSet, the user will see all replicas as currentReplicas labeled with the currentRevision. If the user gets the Pods in their set, the version of the software that they initially deployed (say MySQL 5.7.17) will be identical for all of these Pods and correspond to the currentRevision of the StatefulSet as communicated by its status. If the user gets the StatefulSet and all of the Pods in the StatefulSet, they will see that revision labels of the Pods match the currentRevision of the StatefulSet. They can correlate this back to the product they initially deployed.

When the user updates the StatefulSet, say to deploy MySQL 5.7.18, while the update is progress all Pods running 5.7.17 are labeled with currentRevision, all Pods with MySQL 5.7.18 are labeled with updateRevision, and the current and updated replicas provide accounting for the number of Pods with each version. This models the users intention. The current version that they successfully deployed is 5.7.17 and update they wish to deploy 5.7.18. If the user gets the StatefulSet and its Pods, they can clearly distinguish which Pods correspond to which versions by their labeling.

When the update completes, the current version of the software is 5.7.18, the currentRevision of the StatefulSet is set to the revision (previously updateRevision) that corresponds to this software, and the replica counters are updated to reflect the fact that all Pods in the StatefulSet are at this version. This communicates that the current state of the system is at the state that the user desired when applying the update. If the user gets the StatefulSet and all of its generated Pods, they will see that their desired state has been realized, the current revision of the StatefulSet is the one that they applied. All Pods in the StatefulSet are labeled with this revision, and all are running MySQL 5.7.18.

If I understand the approach that you are suggesting, I think that we will send the user false (or at least confusing) signals. Consider that, in the same scenario as above, when user initially deploys the StatefulSet the the currentRevision will correspond to MySQL 5.7.17 as above, and when the update completes the currentRevision will still correspond to MySQL 5.7.17. The updateRevision will correspond to 5.7.18. Even though the update is complete, if the user gets the StatefulSet and all of its Pods, they will see that no Pods correspond to the currentRevision. This seems very counter-intuitive to me. Whenever I update software or configuration (in any context), if the update is successful, I expect the system to report that current version of the software is the version that I applied. If we take this, approach we signal to the user that the current version of the software is the prior version, and the update version is the version that they applied.
Also, if the user updates the StatefulSet again, the currentRevision will switch to the updateRevision at the same time the updateRevision is updated to indicate a new version of the product is being deployed. So we will eventually signal that the currentRevision of the StatefulSet matches the update the user previously applied, but we will only do so when they perform another update. This indicates that the updateRevision only becomes current when a new update is performed.
This approach would mean that the currentRevision is current version during initial deployment, but after the first update, the currentRevision is really the previous version, and the update revision is the currentRevision. The exception to this is when an update is actually in progress. Only then would the current and update revisions accurately model the users intentions.

kargakis replied:
> It seems counter intuitive to me that when an update completes successfully that the reported current version of the deployed product would be inconsistent with the version newly deployed product.

Not sure I am proposing such a thing. Let me walk through an actual example I am thinking:

--> statefulset "foo" is deployed (stable state)
sts.spec.replicas = 3
sts.status.replicas = 3
sts.status.readyReplicas = 3
sts.status.currentReplicas = 3
sts.status.updatedReplicas = 3
sts.status.currentRevision = foo-1
sts.status.updatedRevision = foo-1

--> statefulset "foo" update applied
sts.spec.replicas = 3
sts.status.replicas = 3
sts.status.readyReplicas = 3
sts.status.currentReplicas = 3
sts.status.updatedReplicas = 0
sts.status.currentRevision = foo-1
sts.status.updatedRevision = foo-2

--> statefulset "foo" scaled up/down
sts.spec.replicas = 3
sts.status.replicas = 3
sts.status.readyReplicas = 3
sts.status.currentReplicas = 2
sts.status.updatedReplicas = 1
sts.status.currentRevision = foo-1
sts.status.updatedRevision = foo-2

--> statefulset "foo" scaled up/down
sts.spec.replicas = 3
sts.status.replicas = 3
sts.status.readyReplicas = 3
sts.status.currentReplicas = 1
sts.status.updatedReplicas = 2
sts.status.currentRevision = foo-1
sts.status.updatedRevision = foo-2

--> statefulset "foo" scaled up/down
sts.spec.replicas = 3
sts.status.replicas = 3
sts.status.readyReplicas = 3
sts.status.currentReplicas = 0
sts.status.updatedReplicas = 3
sts.status.currentRevision = foo-1
sts.status.updatedRevision = foo-2

--> statefulset "foo" update complete (stable state)
sts.spec.replicas = 3
sts.status.replicas = 3
sts.status.readyReplicas = 3
sts.status.currentReplicas = 3
sts.status.updatedReplicas = 3
sts.status.currentRevision = foo-2
sts.status.updatedRevision = foo-2

In stable state, it's obvious that the current revision is the updated (latest) revision and all the current replicas are updated. All of the status API fields comply to their documentation.

kow3ns replied:
I see. This makes sense, and if this is the current behavior for Daemonset I can agree to it. It does differ from the design though. So do we want to implement this in this PR, or modify the design and implement the modification in a subsequent PR?

kargakis replied:
I thought that the only difference between what you had and what I am suggesting are the values in the updatedRevision and updatedReplicas fields so the difference in design is not that big. Yeah, both DS/D are using an UpdatedReplicas field (UpdatedNumberScheduled in DS) that already work as the example above.

kow3ns replied:
It is not a huge difference.

kow3ns commented:
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

krmayankk commented on pkg/apis/apps/v1beta1/types_swagger_doc_generated.go:
```diff
@@ -220,12 +231,27 @@ func (StatefulSetSpec) SwaggerDoc() map[string]string {
+var map_StatefulSetUpdateStrategy = map[string]string{
+	"":          "StatefulSetUpdateStrategy indicates the strategy that the StatefulSet controller will use to perform updates. It includes any additional parameters necessary to perform the update for the indicated strategy.",
+	"type":      "Type indicates the type of the StatefulSetUpdateStrategy.",
+	"partition": "Partition is used to communicate the ordinal at which to partition the StatefulSet when Type is PartitionStatefulSetStrategyType. This value must be set when Type is PartitionStatefulSetStrategyType, and it must be nil otherwise.",
```
Valid values include 1 ---(replicas -1)

kow3ns replied:
@krmayankk this file is generated, but I get your meaning. However, you should add review to the comments in types.go so they are actionable.

krmayankk commented on pkg/controller/history/controller_history.go:
```diff
@@ -0,0 +1,490 @@
+	var owned []*apps.ControllerRevision
+	for i := range history {
+		ref := controller.GetControllerOf(history[i])
+		if ref == nil || ref.UID == parent.GetUID() {
```
if we dont find an owner reference, we shouldnt count that history as owner

kow3ns replied:
if the reference is nil no controller owns it. It is added to owned. If no controller owns it we attempt to adopt it.

krmayankk commented on pkg/controller/statefulset/stateful_set_control.go:
```diff
@@ -57,20 +71,220 @@ type defaultStatefulSetControl struct {
+		return nil
+	}
+	// delete any non-live history to maintain the revision limit.
+	history = history[:int(*set.Spec.RevisionHistoryLimit)]
```
origiHistory = [0 1 2 3 4 ]
revisionLimit  = 3
newHistory = origHistory[ :3 ]
newHistory = [0 1 2] which is deleted in the subsequent loop. so the remaining left is 3 and 4. What we really want is 0 and 1 oldest revision deleted and the only remaining should be [2,3 ,4]
so history = history[ : len(history)-revisionLimit]  @kow3ns ?

kow3ns commented:
@kargakis I updated the status reporting as per your suggestion. currentRevision == updateRevision is just just as good as updateRevision == "".

kow3ns commented:
@k8s-bot pull-kubernetes-unit test this

kargakis commented on pkg/controller/statefulset/stateful_set_control.go:
```diff
@@ -57,20 +71,222 @@ type defaultStatefulSetControl struct {
+// all Pods with ordinal less than UpdateStrategy.Partition.Ordinal must be at Status.CurrentRevision and all other
+// Pods must be at Status.UpdateRevision. If the returned error is nil, the returned StatefulSetStatus is valid and the
+// update must be recorded. If the error is not nil, the method should be retried until successful.
+func (ssc *defaultStatefulSetControl) updateStatefulSet(set *apps.StatefulSet,
```
nit: Suggestion for better readability
```go
func (ssc *defaultStatefulSetControl) updateStatefulSet(
  set *apps.StatefulSet,
  currentRevision *apps.ControllerRevision,
  updateRevision *apps.ControllerRevision,
  pods []*v1.Pod,
) (*apps.StatefulSetStatus, error) {
```

kargakis commented on pkg/controller/statefulset/stateful_set_test.go:
```diff
@@ -171,6 +172,7 @@ func TestStatefulSetControllerBlocksScaling(t *testing.T) {
 		t.Error(err)
 	}
 	if len(pods) != 3 {
+
```
kill this newline

kargakis commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -255,6 +281,122 @@ func newStatefulSetPod(set *apps.StatefulSet, ordinal int) *v1.Pod {
+// current revision. updateSet is the representation of the set at the updateRevision. currentRevision is the name of
+// the current revision. updateRevision is the name of the update revision. ordinal is the ordinal of the Pod. If the
+// returned error is nil, the returned Pod is valid.
+func newVersionedStatefulSetPod(currentSet *apps.StatefulSet, updateSet *apps.StatefulSet, currentRevision string, updateRevision string, ordinal int) *v1.Pod {
```
nit: `currentSet, updateSet *apps.StatefulSet, currentRevision, updateRevision string, ordinal int`

kargakis commented on pkg/kubectl/rollout_status.go:
```diff
@@ -107,3 +114,34 @@ func (s *DaemonSetStatusViewer) Status(namespace, name string, revision int64) (
+		return fmt.Sprintf("partitioned roll out complete: %d new pods have been updated...\n",
+			sts.Status.UpdatedReplicas), false, nil
+	}
+	if sts.Status.UpdateRevision != "" {
```
You probably want to update this one too as part of moving away from resetting updateRevision and updatedReplicas.

kargakis commented on pkg/kubectl/rollout_status_test.go:
```diff
@@ -19,7 +19,10 @@ package kubectl
 import (
 	"testing"
 
+	"fmt"
```
collapse with rest of the stdlib imports

kargakis commented:
You need to fix `rollout status` then this LGTM

kargakis commented:
Also, when you squash, can you keep at least three separate commits: generated changes, kubectl changes, rest of the changes?

krmayankk commented on pkg/controller/statefulset/stateful_set_control.go:
```diff
@@ -57,20 +71,222 @@ type defaultStatefulSetControl struct {
+	}
+	historyLen := len(history)
+	historyLimit := int(*set.Spec.RevisionHistoryLimit)
+	if historyLen < historyLimit {
```
nit: <=

krmayankk commented on pkg/controller/statefulset/stateful_set_control.go:
```diff
@@ -83,45 +299,49 @@ func (ssc *defaultStatefulSetControl) UpdateStatefulSet(set *apps.StatefulSet, p
+	// sort the condemned Pods by their ordinals
+	sort.Sort(ascendingOrdinal(condemned))
+
+	// find the first unhealthy Pod
```
nit: why do we care about the firstUnhealthyPod, when you get time, may be add a comment in subsequent PR

kow3ns replied:
The are comments above where each check against the variable is made

krmayankk commented on pkg/controller/statefulset/stateful_set_control.go:
```diff
@@ -168,31 +417,119 @@ func (ssc *defaultStatefulSetControl) UpdateStatefulSet(set *apps.StatefulSet, p
+			continue
+		}
+		// if we are in monotonic mode and the condemned target is not the first unhealthy Pod block
+		if !isRunningAndReady(condemned[target]) && monotonic && condemned[target] != firstUnhealthyPod {
```
nit: add comment about the special handling of firstUnhealthPod

krmayankk replied:
is this checking saying that before we can delete or scale down, the extra pods must be healthy and running ? Is it for safety of clustered applications ?

kow3ns replied:
Yes. As documented, the StatefulSet guarantees for monotonic mode (the original mode) is that the predecessors (wrt ordinal) of every Pod are running and ready before a new Pod is created.

krmayankk reviewed (approved):
LGTM some nit comments

kow3ns commented:
As @smarterclayton pointed out we need to add additional e2e in a followup #46942

Commit:
Implements StatefulSet update
Implements history utilities for ControllerRevision in the controller/history package
StatefulSetStatus now has additional fields for consistency with DaemonSet and Deployment
StatefulSetStatus.Replicas now represents the current number of createdPods and StatefulSetStatus.ReadyReplicas is the current number of ready Pods

Commit:
Implements kubectl rollout status and history for StatefulSet

Commit:
Auto generated code for StatefulSet update

kow3ns commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

kargakis commented:
/lgtm

kargakis commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

tnozicka commented:
> the finalizer did not appear to work. It doesn't even pass unit tests so I pulled it.

@kow3ns was it before https://github.com/kubernetes/kubernetes/pull/46500 merged? I can't see the old changes but that has been an issue with the finalizer for the referenced PR https://github.com/kubernetes/kubernetes/pull/45764

smarterclayton commented:
flake in node authorizer, jordan is looking at it:

@k8s-bot pull-kubernetes-unit test this

kow3ns commented:
@k8s-bot pull-kubernetes-e2e-kops-aws test this

kow3ns commented:
@k8s-bot pull-kubernetes-e2e-kops-aws test this

mikegrass commented on pkg/controller/history/controller_history.go:
```diff
@@ -0,0 +1,490 @@
+	return cr, nil
+}
+
+// HashControllerRevision hashes the contents of revision's Data using FNV hashing. If probe is not nil, the byte value
```
Nit: comment says "the byte value" is added to the hash, but it's the string representation of `*probe` that gets written.

Do you want to use `encoding/binary` and do `binary.Write(hf, binary.LittleEndian, *probe)` instead?

mikegrass commented on pkg/controller/history/controller_history.go:
```diff
@@ -0,0 +1,490 @@
+		return nil, err
+	}
+	clone := any.(*apps.ControllerRevision)
+
```
fakeHistory's CreateControllerRevision sets the clone's namespace to the parent's namespace. Do you need to do that here?

mikegrass replied:
As an aside, fakeHistory's CreateControllerRevision and ListControllerRevisions are nearly identical to realHistory's implementations. Opportunity for factoring out the common code?

mikegrass commented on pkg/controller/history/controller_history.go:
```diff
@@ -0,0 +1,490 @@
+}
+
+func (rh *realHistory) ReleaseControllerRevision(parent metav1.Object, revision *apps.ControllerRevision) (*apps.ControllerRevision, error) {
+	// Use strategic merge patch to add an owner reference indicating a controller ref
```
Nit: s/add/delete/

mikegrass reviewed (commented):
Some minor issues. Overall looks good to me.

kargakis commented on pkg/api/testing/fuzzer.go:
```diff
@@ -723,6 +723,13 @@ func appsFuncs(t apitesting.TestingCommon) []interface{} {
 				s.Spec.PodManagementPolicy = apps.OrderedReadyPodManagement
 			}
+			if len(s.Spec.UpdateStrategy.Type) == 0 {
+				s.Spec.UpdateStrategy.Type = apps.RollingUpdateStatefulSetStrategyType
```
Should you also update this one to OnDelete?

kargakis commented on pkg/controller/statefulset/stateful_set.go:
```diff
@@ -95,8 +97,9 @@ func NewStatefulSetController(
+			NewRealStatefulSetStatusUpdater(kubeClient, setInformer.Lister()),
+			history.NewHistory(kubeClient, revInformer.Lister()),
 		),
 		pvcListerSynced: pvcInformer.Informer().HasSynced,
```
Since you are using the controllerRevision cache to list CRs, you should wait for it to be synced before starting the sts controller. Means you need a crListerSynced field which you then pass in controller.WaitForCacheSync (L150).

wojtek-t commented:
/approve no-issue (there is issue from features repo, but bot still doesn't understand it).

wojtek-t commented:
/approve no-issue

kargakis commented:
> there is issue from features repo, but bot still doesn't understand it

There is https://github.com/kubernetes/features/issues/188

janetkuo commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -255,6 +281,122 @@ func newStatefulSetPod(set *apps.StatefulSet, ordinal int) *v1.Pod {
+		return nil, err
+	}
+	var raw map[string]interface{}
+	json.Unmarshal([]byte(str), &raw)
```
Would you want to check the returned error?

janetkuo commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -255,6 +281,122 @@ func newStatefulSetPod(set *apps.StatefulSet, ordinal int) *v1.Pod {
+// PodSpecTemplate. We can modify this later to encompass more state (or less) and remain compatible with previously
+// recorded patches.
+func getPatch(set *apps.StatefulSet) ([]byte, error) {
+	str, err := runtime.Encode(patchCodec, set)
```
Remove `patchCodec` and use `json.Marshal` here instead, csi team wants to avoid using api.Codecs as much as possible https://github.com/kubernetes/kubernetes/pull/47075#discussion_r120467218

janetkuo commented:
LGTM, comments can be fixed in follow-up PRs

janetkuo commented:
/lgtm
